File: dev_stt_2019/embedded_system_surprisy/embedded_system/library_py/imu_py/Imu.py
# import sys
# sys.path.append('.')
import RTIMU, json, time, math
import threading

class Imu(threading.Thread):
    def __init__(self,code="Surprice"):
        print("Loading Imu Sensor...");
        threading.Thread.__init__(self)
        self.status = False;
        self.statusPressure = False;
        self.pressure = 1;
        self.data = {}
        self.fusionPose = {'x':0,'y':0,'z':0}

        SETTINGS_FILE = "settings/RTIMULib"
        if(code == "Hector"):
            SETTINGS_FILE = SETTINGS_FILE + "_" + code
            print(SETTINGS_FILE);
        self.s = RTIMU.Settings(SETTINGS_FILE)
        self.imu = RTIMU.RTIMU(self.s)
        self.rtpressure = RTIMU.RTPressure(self.s)

        self.imu.IMUInit()
        # if (not self.imu.IMUInit()):
            # print("IMU Init Failed")
            # sys.exit(1)
        # else:
            # print("IMU Init Succeeded");

        self.imu.setSlerpPower(0.02)
        self.imu.setGyroEnable(True)
        self.imu.setAccelEnable(True)
        self.imu.setCompassEnable(True)

        self.rtpressure.pressureInit()
        # if (not self.rtpressure.pressureInit()):
        #     sys.exit(1)
        # # else:
        # #     print("Pressure sensor Init Succeeded")

        self.poll_interval = self.imu.IMUGetPollInterval()
        print("Loaded Imu Sensor");
    def getData(self):
        return self.data
    def getDataFusion(self):
        return self.fusionPose
    def getDataJson(self):
        return json.dumps(self.data)

    def getPressure(self):
        return self.pressure
    def getAltitude(self,P0):
        #P0
        #T0
        #z0 da prendere da getData prima della function
        g=9.81;
        R=287;
        T0=290;
        P=self.pressure;
        if(self.pressure == 0):
            P = 1.
        return (-(R*T0)/g)*math.log(P/P0);
    def uploadData(self):
        flag = True;
        def computeHeight(pressure):
            return 44330.8 * (1 - pow(pressure / 1013.25, 0.190263));
        while flag:
            if self.imu.IMURead():
                data = self.imu.getIMUData()
                (data["pressureValid"], data["pressure"], data["temperatureValid"], data["temperature"]) = self.rtpressure.pressureRead()

                if data["pressureValid"] != 0 and data["temperatureValid"] != 0:
                    pressure = data['pressure']
                    data["compute_height"] = computeHeight(pressure);
                    flag = False;
                self.data = data;
                # print(math.sqrt(math.pow(self.data['compass'][0],2)+math.pow(self.data['compass'][1],2)+math.pow(self.data['compass'][2],2)));
                self.fusionPose['x'] = self.data['fusionPose'][2]*(180/3.141);
                self.fusionPose['y'] = self.data['fusionPose'][2]*(180/3.141);
                self.fusionPose['z'] = self.data['fusionPose'][2]*(180/3.141);
    def run(self):
        def computeHeight(pressure):
            return 44330.8 * (1 - pow(pressure / 1013.25, 0.190263));
        while True:
            try:
                if self.imu.IMURead():
                    data = self.imu.getIMUData()
                    # print(data['compass'][0])
                    (data["pressureValid"], data["pressure"], data["temperatureValid"], data["temperature"]) = self.rtpressure.pressureRead()
                    if data["pressureValid"] != 0 and data["temperatureValid"] != 0:
                        self.data = data;
                        self.status = True;
                    else:
                        self.status = False;
                    #     self.pressure = data['pressure']
                    #     data["compute_height"] = computeHeight(self.pressure);
                    #     self.fusionPose['x'] = self.data['fusionPose'][2]*(180/3.141);
                    #     self.fusionPose['y'] = self.data['fusionPose'][2]*(180/3.141);
                    #     self.fusionPose['z'] = self.data['fusionPose'][2]*(180/3.141);
                    # print(self.pressure);
                else:
                    self.status = False;
                time.sleep(0.001)
            except KeyboardInterrupt:
                self.status = False;
                time.sleep(3)

File: dev_stt_2019/embedded_system_surprisy/embedded_system/library_cpp/imu_cpp/Imu.cpp
#include "Imu.h"


Imu::Imu() {
    RTIMUSettings *settings = new RTIMUSettings("RTIMULib");

    RTimu = RTIMU::createIMU(settings);
    RTpressure = RTPressure::createPressure(settings);
    if ((RTimu == NULL) || (RTimu->IMUType() == RTIMU_TYPE_NULL)) {
        printf("No IMU found\n");
        exit(1);
    }
    RTimu->IMUInit();

    //  this is a convenient place to change fusion parameters

    RTimu->setSlerpPower(0.02);
    RTimu->setGyroEnable(true);
    RTimu->setAccelEnable(true);
    RTimu->setCompassEnable(true);

    //  set up pressure sensor

    if (RTpressure != NULL)
        RTpressure->pressureInit();
}
void Imu::updateData() {
    // usleep(RTimu->IMUGetPollInterval() * 1000);
    if (RTimu->IMURead()) {
        RTIMU_DATA RTimuData = RTimu->getIMUData();
        if (RTpressure != NULL)
            RTpressure->pressureRead(RTimuData);

        printf("Sample rate : %f\n", RTimuData.fusionPose.z()*RTMATH_RAD_TO_DEGREE);
        fusionPoseZ = RTimuData.fusionPose.z()*RTMATH_RAD_TO_DEGREE;
        if (RTpressure != NULL) {
            pressure = RTimuData.pressure;
            temperature = RTimuData.temperature;
            // printf("Pressure: %4.1f, height above sea level: %4.1f, temperature: %4.1f\n",
                   // RTimuData.pressure, RTMath::convertPressureToHeight(RTimuData.pressure), RTimuData.temperature);
        }

    }
}

void Imu::operator()() {
    while(1){
        if (RTimu->IMURead()) {
            RTIMU_DATA RTimuData = RTimu->getIMUData();
            if (RTpressure != NULL)
                RTpressure->pressureRead(RTimuData);

            // printf("Sample rate : %f\n", RTimuData.fusionPose.z()*RTMATH_RAD_TO_DEGREE);
            fusionPoseZ = RTimuData.fusionPose.z()*RTMATH_RAD_TO_DEGREE;
            if (RTpressure != NULL) {
                pressure = RTimuData.pressure;
                temperature = RTimuData.temperature;
                // printf("Pressure: %4.1f, height above sea level: %4.1f, temperature: %4.1f\n",
                       // RTimuData.pressure, RTMath::convertPressureToHeight(RTimuData.pressure), RTimuData.temperature);
            }
        }
        usleep(100000);

    }
};
double Imu::getPressure() {
    return pressure;
}
double Imu::getTemperature() {
    return temperature;
}
double Imu::getFusionPoseZ() {
    return fusionPoseZ;
}

void Imu::greeting(){
    int k = 0;
    while(true){
        if (RTimu->IMURead()) {
            RTIMU_DATA RTimuData = RTimu->getIMUData();
            if (RTpressure != NULL)
                RTpressure->pressureRead(RTimuData);

            printf("Sample rate : %f\n", RTimuData.fusionPose.z()*RTMATH_RAD_TO_DEGREE);
            fusionPoseZ = RTimuData.fusionPose.z()*RTMATH_RAD_TO_DEGREE;
            if (RTpressure != NULL) {
                pressure = RTimuData.pressure;
                temperature = RTimuData.temperature;
                // printf("Pressure: %4.1f, height above sea level: %4.1f, temperature: %4.1f\n",
                       // RTimuData.pressure, RTMath::convertPressureToHeight(RTimuData.pressure), RTimuData.temperature);
            }
            printf("_%d",k);
            k++;
        }
        usleep(10000);

    }
}
